### PR TITLE
Add support for openshift-logging tenant mode

### PR DESCRIPTION
**Pull request by periklis**
This PR implements the tenant mode `openshift-logging` for fully-automated configuration management for the `lokistack-gateway`. The change set adds the following functionality:
- Generate OIDIC secrets for each tenant
- Augments the `lokistack-gateway` deployment by a new sidecar `opa-openshift` to handle authorization to the api server.
- Augments the `lokistack-gateway` configuration by generating sane defaults for `tenants.Authentication` and `tenants.Authorization` blocks.
- Augment the `lokistack-gateway` service for `opa-openshift` sidecar to scrape metrics.

### CODE REVIEW

File: internal/manifests/internal/gateway/gateway-tenants.yaml [hunk 1]
```diff
@@ -17,24 +17,22 @@ tenants:
     issuerCAPath: {{ $secret.IssuerCAPath }}
     {{- end -}}
     {{- end -}}
-    {{- end -}}
-    {{ print "\n" }}
```
Comment (sasagarw) [inline internal/manifests/internal/gateway/gateway-tenants.yaml:21]: Was it working without `{{ print "\n" }}`? It was generating wrong yaml previously when we omit this.

Comment (periklis) [reply internal/manifests/internal/gateway/gateway-tenants.yaml:21]: Yes, you need to tune when to use the `-` in starting blocks `{{` and end blocks `}}`.

File: internal/manifests/gateway.go [hunk 2]
```diff
@@ -269,54 +279,54 @@ func gatewayConfigOptions(opt Options) gateway.Options {
...
+				MountPath: gateway.LokiGatewayTLSDir,
 			},
 		},
 		Args: []string{
-			fmt.Sprintf("--tls.internal.server.cert-file=%s", path.Join(gateway.LokiGatewayTLSDir, "cert")),
-			fmt.Sprintf("--tls.internal.server.key-file=%s", path.Join(gateway.LokiGatewayTLSDir, "key")),
-			fmt.Sprintf("--tls.healthchecks.server-ca-file=%s", path.Join(gateway.LokiGatewayTLSDir, "ca")),
+			fmt.Sprintf("--tls.internal.server.cert-file=%s", certFile),
+			fmt.Sprintf("--tls.internal.server.key-file=%s", keyFile),
+			// fmt.Sprintf("--tls.healthchecks.server-ca-file=%s", caFile),
```
Comment (sasagarw) [inline internal/manifests/gateway.go:329]: @periklis Any reason why we have commented out things related to `server-ca-file` everywhere?

Comment (periklis) [reply internal/manifests/gateway.go:329]: @sasagarw The CA file is something we need to implement for TLS metrics. This requires a separate controller to reconcile for the TrustedCABundle in OCP as we similarly do in EO and CLO for Kibana. Currently the pods don't start with these options.

File: internal/handlers/internal/gateway/tenant_secrets.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	case lokiv1beta1.Static, lokiv1beta1.Dynamic:
+		return extractUserProvided(ctx, k, req, stack)
+	case lokiv1beta1.OpenshiftLogging:
+		return createOpenShiftLoggingSecrets(ctx, k, req, scheme, stack)
+	}
+
+	return nil, nil
+}
+
+func extractUserProvided(
```
Comment (sasagarw) [inline internal/handlers/internal/gateway/tenant_secrets.go:44]:
```suggestion
func extractUserProvidedSecrets(
```
Adding `Secrets` to the function name will make it more understandable that we are trying to extract user provided secrets in this function. WDYT?

File: internal/manifests/gateway_tenant_sidecars_test.go [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+	lokiv1beta1 "github.com/ViaQ/loki-operator/api/v1beta1"
+	"github.com/ViaQ/loki-operator/internal/manifests/internal/gateway"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func Test_ApplyGatewayDefaultsOptions(t *testing.T) {
```
Comment (sasagarw) [inline internal/manifests/gateway_tenant_sidecars_test.go:17]:
```suggestion
func TestApplyGatewayDefaultsOptions(t *testing.T) {
```
For consistency in naming everywhere.

File: internal/handlers/internal/gateway/tenant_secrets_test.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var scheme = runtime.NewScheme()
+
+func Test_GetTenantSecrets_StaticMode(t *testing.T) {
```
Comment (sasagarw) [inline internal/handlers/internal/gateway/tenant_secrets_test.go:25]:
```suggestion
func TestGetTenantSecrets_StaticMode(t *testing.T) {
```
For consistency in naming everywhere.

File: internal/handlers/internal/gateway/tenant_secrets_test.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+			TenantName:   "test",
+			ClientID:     "test",
+			ClientSecret: "test",
+			IssuerCAPath: "/path/to/ca/file",
+		},
+	}
+	require.ElementsMatch(t, ts, expected)
+}
+
+func Test_GetTenantSecrets_DynamicMode(t *testing.T) {
```
Comment (sasagarw) [inline internal/handlers/internal/gateway/tenant_secrets_test.go:88]:
```suggestion
func TestGetTenantSecrets_DynamicMode(t *testing.T) {
```
For consistency in naming everywhere.

File: internal/handlers/internal/gateway/tenant_secrets_test.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+			TenantName:   "test",
+			ClientID:     "test",
+			ClientSecret: "test",
+			IssuerCAPath: "/path/to/ca/file",
+		},
+	}
+	require.ElementsMatch(t, ts, expected)
+}
+
+func Test_GetTenantSecrets_OpenShiftLoggingMode(t *testing.T) {
```
Comment (sasagarw) [inline internal/handlers/internal/gateway/tenant_secrets_test.go:151]:
```suggestion
func TestGetTenantSecrets_OpenShiftLoggingMode(t *testing.T) {
```
For consistency in naming everywhere.

**Review by sasagarw — COMMENTED** [review]
Just small changes otherwise LGTM :rocket::rocket:

**Review by sasagarw — APPROVED** [review]
LGTM

**Pull request merged**